### PR TITLE
NMS-13362: Add Kafka consumer for OpenNMS events

### PR DESCRIPTION
**Pull request by cgorantla**
Add new  feature `opennms-kafka-consumer` that allows to consume events from Kafka.
This allows OpenNMS to consume events that are generated when OpenNMS is Offline.

Also it simplifies the event generation to be protobuf.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13362

### CODE REVIEW

File: features/kafka/consumer/src/main/proto/kafka-consumer-events.proto [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    MAJOR = 5;
+    CRITICAL = 6;
+}
+
+message EventParameter {
+    string name = 1;
+    string value = 2;
+}
+
+message Event {
```
Comment (cgorantla) [inline features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:22]: @j-white @RangerRick   I brought most of the elements in `org.opennms.netmgt.xml.event.Event` to proto.
Let me know if this looks fine.

**Comment by cgorantla** [comment]
Currently this is targeting to `develop` but think this can be backported to `foundation-2020`

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 2]
```diff
@@ -0,0 +1,50 @@
+
+[[ga-events-sources-kafka]]
+
+= Kafka Consumer (Events)
+
+Events can also be created by routing them to a specific topic on Kafka.
+The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.
+
+Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]
+See `kafka-consumer-events.proto`  for the model definitions.
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:10]:
```suggestion
See `kafka-consumer-events.proto` for the model definitions.
```

File: features/kafka/consumer/src/main/proto/kafka-consumer-events.proto [hunk 3]
```diff
@@ -0,0 +1,37 @@
+syntax = "proto3";
+option java_package = "org.opennms.features.kafka.consumer.events";
+option java_outer_classname = "EventsProto";
+
+// The values differ from the standard codes in OpenNMS
+// since proto3 enforces us to start at 0
```
Comment (RangerRick) [inline features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: Oof... I don't suppose we could just make 0 be "unknown" or something, and have it not be used? kinda sucks to have them off-by-one.

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+== Enable Kafka Consumer
+
+The Kafka Consumer is disabled by default and can be enabled as follows.
+
+First, log in to the Karaf shell of your {page-component-title} instance and configure the Kafka client settings to point to your Kafka broker.
+See link:https://kafka.apache.org/27/documentation.html#consumerconfigs[Consumer Configs] for a complete list of available options.
+
+[source]
+----
+$ ssh -p 8101 admin@localhost
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:23]:
```suggestion
ssh -p 8101 admin@localhost
```

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+...
+admin@opennms()> config:edit org.opennms.features.kafka.consumer.client
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:25]:
```suggestion
config:edit org.opennms.features.kafka.consumer.client
```

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+admin@opennms()> config:property-set bootstrap.servers 127.0.0.1:9092
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:26]:
```suggestion
config:property-set bootstrap.servers 127.0.0.1:9092
```

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+admin@opennms()> config:update
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:27]:
```suggestion
config:update
```

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 6 above)
Comment (indigo423) [reply docs/modules/operation/pages/events/sources/kafka.adoc:26]: We should probably add a callout for the 127.0.0.1 bootstrap server and port like this:

> Connect to the following Kafka nodes and adjust the IPs or FQDNs with the Kafka port (9092) accordingly.

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+----
+
+Next, install the `opennms-kafka-consumer` feature from that same shell using:
+
+[source]
+----
+admin@opennms()> feature:install opennms-kafka-consumer
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:34]:
```suggestion
feature:install opennms-kafka-consumer
```

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+The Kafka Consumer by default consumes events from topic named `opennms-events` and forward them to `Eventd`.
+
+== Enable Kafka Consumer
+
+The Kafka Consumer is disabled by default and can be enabled as follows.
+
+First, log in to the Karaf shell of your {page-component-title} instance and configure the Kafka client settings to point to your Kafka broker.
+See link:https://kafka.apache.org/27/documentation.html#consumerconfigs[Consumer Configs] for a complete list of available options.
+
+[source]
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:21]:
```suggestion
[source, console]
```

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 4 above)
Comment (indigo423) [reply docs/modules/operation/pages/events/sources/kafka.adoc:23]: We should split the source sections to indicate what is console and what is Karaf shell like this:

```
.Configure the Kafka features and the Minion location via the Karaf Shell
[source, console]
----
ssh -p 8201 admin@localhost
----

.Configure the Event consumer client
[source, karaf]
----
config:edit org.opennms.features.kafka.consumer.client
config:property-set bootstrap.servers 127.0.0.1:9092
config:update
----
```

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+----
+admin@opennms()> feature:install opennms-kafka-consumer
+----
+
+To ensure that the feature continues to be installed on subsequent restarts, add `opennms-kafka-consumer` to the `featuresBoot` property in the `$\{OPENNMS_HOME}/etc/org.apache.karaf.features.cfg`.
+
+== Configure Topic Name
+
+.Karaf login, configuration of events topic
+[source]
```
Comment (indigo423) [inline docs/modules/operation/pages/events/sources/kafka.adoc:42]: Same as above, split and no prompts so we can use copy & paste.

**Review by indigo423 — CHANGES_REQUESTED** [review]
Just minor things so terminal in and outputs are consistent.

File: features/kafka/consumer/src/main/proto/kafka-consumer-events.proto (hunk 3 above)
Comment (Naicisum) [reply features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: Why not do the industry standard, use `UNUSED`, `PAD0`, `PAD1`, or `RESERVED`?

**Comment by Naicisum** [comment]
Probably need to code a runtime config check to ensure a loop between the kafka-producer event topic if configured, and this feature's event topic does not occur.

**Comment by opennms-bamboo** [comment]
Welcome CI/CD System (cicd-system@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username CI/CD System (cicd-system@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: features/kafka/consumer/src/main/proto/kafka-consumer-events.proto (hunk 3 above)
Comment (cgorantla) [reply features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: `INDETERMINATE` seems to be 0 so there is already `unknown` case.  Mapping seems to be easier by matching label.

Comment (cgorantla) [reply features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: @Naicisum I think this issue is mostly trying to use our `Event` and match it with protobuf and get events consumed directly on OpenNMS.

**Comment by cgorantla** [comment]
> Probably need to code a runtime config check to ensure a loop between the kafka-producer event topic if configured, and this feature's event topic does not occur.

That's good point. We may need to check if topic currently exist before subscribing and we have to do that in lot of places. For now I renamed it as `opennms-kafka-events` and added a note on doc.

**Comment by cgorantla** [comment]
@indigo423 Thanks for the review. Handled them. Please take a look again.

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 11]
```diff
@@ -59,3 +59,4 @@ config:property-set eventsTopic opennms-kafka-events
 config:update
 ----
 
+NOTE: While configuring events topic, make sure that it doesn't conflict with other topics in OpenNMS subsystem. If you are unsure, keep the default one.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/events/sources/kafka.adoc:62]: NOTE: While configuring the eventsTopic, make sure that it doesn't conflict with other topics in the OpenNMS subsystem. If you are unsure, keep the default one.

Comment (Bonrob2) [reply docs/modules/operation/pages/events/sources/kafka.adoc:62]:
```suggestion
NOTE: While configuring the eventsTopic, make sure that it doesn't conflict with other topics in the OpenNMS subsystem. If you are unsure, keep the default one.
```

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 12]
```diff
@@ -0,0 +1,50 @@
+
+[[ga-events-sources-kafka]]
+
+= Kafka Consumer (Events)
+
+Events can also be created by routing them to a specific topic on Kafka.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/events/sources/kafka.adoc:6]: You can also create events by routing them to a specific topic on Kafka.

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+The Kafka Consumer feature allows events to be consumed from specific topic on Kafka.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/events/sources/kafka.adoc:7]: The Kafka Consumer feature allows events to be consumed from a specific topic on Kafka.

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+Events should be created in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)]
```
Comment (Bonrob2) [inline docs/modules/operation/pages/events/sources/kafka.adoc:9]: Create events in protobuf fomrat link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)].

File: docs/modules/operation/pages/events/sources/kafka.adoc [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+See `kafka-consumer-events.proto`  for the model definitions.
+
+The Kafka Consumer by default consumes events from topic named `opennms-events` and forward them to `Eventd`.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/events/sources/kafka.adoc:12]: The Kafka Consumer by default consumes events from topics named `opennms-events` and forwards them to `Eventd`.

**Review by Bonrob2 — CHANGES_REQUESTED** [review]

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 12 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/events/sources/kafka.adoc:6]:
```suggestion
You can also create events by routing them to a specific topic on Kafka.
```

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 13 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/events/sources/kafka.adoc:7]:
```suggestion
The Kafka Consumer feature allows events to be consumed from a specific topic on Kafka.
```

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 15 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/events/sources/kafka.adoc:12]:
```suggestion
The Kafka Consumer by default consumes events from topics named `opennms-events` and forwards them to `Eventd`.
```

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 14 above)
Comment (Bonrob2) [inline docs/modules/operation/pages/events/sources/kafka.adoc:9]: Events should be created in protobuf format

**Review by Bonrob2 — CHANGES_REQUESTED** [review]

File: docs/modules/operation/pages/events/sources/kafka.adoc (hunk 14 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/events/sources/kafka.adoc:9]:
```suggestion
Create events in protobuf format link:https://developers.google.com/protocol-buffers/[Google Protocol Buffers (GPB)].
```

**Comment by cgorantla** [comment]
@Bonrob2 Thanks the review. Updated.

File: features/kafka/consumer/src/main/proto/kafka-consumer-events.proto (hunk 3 above)
Comment (RangerRick) [reply features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: Sure, but `INDETERMINATE` has a specific meaning/mapping in the OpenNMS code, whereas in this case the 0th entry would be not actually used.  I am not a fan of having 2 different "standards" for what the enums map top (0-indexed and 1-indexed), it makes it the API ambiguous and have different behavior whether you're parsing say, the ReST API vs the Kafka stream.

Comment (j-white) [reply features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: This was copied from here [opennms-kafka-producer.proto#L7](https://github.com/OpenNMS/opennms/blob/opennms-28.0.0-2/features/kafka/producer/src/main/proto/opennms-kafka-producer.proto#L7) for reference - so we already to the start at 0 thing in the producer case.

Comment (RangerRick) [reply features/kafka/consumer/src/main/proto/kafka-consumer-events.proto:6]: oof, alright... then I guess there are [2 standards](https://xkcd.com/927/)  :D

**Comment by Bonrob2** [comment]
LGTM!

**Review by j-white — COMMENTED** [review]
Looks good. Much nicer than XML wrapped in a Protobuf payload :)

**Review by RangerRick — APPROVED** [review]
[ship it squirrel]

**Pull request merged**